commit e46e97ade210e3e960a9a3cf7b3edb651411e598
Author: bdg <1926366608@qq.com>
Date:   Fri Sep 22 17:43:08 2023 +0800

    Return total size of content that has been read on endElement2()
    
    N/A
    
    .

diff --git a/samples/src/MemParse/MemParse.cpp b/samples/src/MemParse/MemParse.cpp
index 61fa66706..6819c7f55 100644
--- a/samples/src/MemParse/MemParse.cpp
+++ b/samples/src/MemParse/MemParse.cpp
@@ -67,7 +67,9 @@
    #endif
 #endif /* ifndef MEMPARSE_ENCODING */
 
-static const char*  gXMLInMemBuf =
+#define MEMPARSE_ENCODING "utf8"
+#if 0
+const char*  gXMLInMemBuf =
 "\
 <?xml version='1.0' encoding='" MEMPARSE_ENCODING "'?>\n\
 <!DOCTYPE company [\n\
@@ -78,13 +80,48 @@ static const char*  gXMLInMemBuf =
 <!ELEMENT developedAt (#PCDATA)>\n\
 ]>\n\n\
 <company>\n\
-    <product>XML4C</product>\n\
+    <product>美丽</product>\n\
     <category idea='great'>XML Parsing Tools</category>\n\
     <developedAt>\n\
       IBM Center for Java Technology, Silicon Valley, Cupertino, CA\n\
     </developedAt>\n\
 </company>\
 ";
+#else
+#if 1
+const char*  gXMLInMemBuf =
+"\
+<?xml version='1.0' encoding='UTF-8' standalone='yes'?>\n\
+<worksheet xmlns='http://schemas.openxmlformats.org/spreadsheetml/2006/main' xmlns:r='http://schemas.openxmlformats.org/officeDocument/2006/relationships' xmlns:xdr='http://schemas.openxmlformats.org/drawingml/2006/spreadsheetDrawing' xmlns:x14='http://schemas.microsoft.com/office/spreadsheetml/2009/9/main' xmlns:mc='http://schemas.openxmlformats.org/markup-compatibility/2006' xmlns:etc='http://www.wps.cn/officeDocument/2017/etCustomData'>\n\
+    <sheetPr>\n\
+        <pageSetUpPr fitToPage='1'/>\n\
+    </sheetPr>\n\
+    <dimension ref='A1:CE100'/>\n\
+    <sheetViews>\n\
+        <sheetView zoomScale='70' zoomScaleNormal='70' zoomScaleSheetLayoutView='50' workbookViewId='0'>\n\
+            <selection activeCell='D47' sqref='A1:CD100'/>\n\
+        </sheetView>\n\
+    </sheetViews>\n\
+    <sheetFormatPr defaultColWidth='9' defaultRowHeight='15.75' customHeight='1'/>\n\
+    <cols>\n\
+        <col min='1' max='1' width='6.16153846153846' style='14' customWidth='1'/>\n\
+        <col min='83' max='16384' width='9' style='11'/>\n\
+    </cols>\n\
+    <sheetData>mmm</sheetData>\n\
+    <legacyDrawing />\n\
+</worksheet>\n\
+";
+#else
+const char*  gXMLInMemBuf =
+"\
+<?xml version='1.0' encoding='UTF-8' standalone='yes'?>\n\
+<worksheet>\n\
+<a/>\n\
+<b>mm</b>\n\
+</worksheet>\n\
+";
+#endif
+#endif
 
 static const char*  gMemBufId = "prodInfo";
 
@@ -253,6 +290,12 @@ int main(int argC, char* argV[])
              << StrX(e.getMessage()) << "\n" << XERCES_STD_QUALIFIER endl;
         errorCode = 4;
     }
+    catch (const UserInterruption& e)
+    {
+        XERCES_STD_QUALIFIER cerr << "\nUserInterruption: " << e.mCode << XERCES_STD_QUALIFIER endl;
+        errorCode = 4;
+    }
+    
     if(errorCode) {
         XMLPlatformUtils::Terminate();
         return errorCode;
diff --git a/samples/src/MemParse/MemParseHandlers.cpp b/samples/src/MemParse/MemParseHandlers.cpp
index ee91ce73b..b76e973c9 100644
--- a/samples/src/MemParse/MemParseHandlers.cpp
+++ b/samples/src/MemParse/MemParseHandlers.cpp
@@ -29,7 +29,78 @@
 #include <xercesc/sax/AttributeList.hpp>
 #include <xercesc/sax/SAXParseException.hpp>
 #include <xercesc/sax/SAXException.hpp>
-
+#include <iostream>
+#include <stdio.h>
+
+extern const char*  gXMLInMemBuf;
+using namespace std;
+
+void XMLChToUtf8(string& utf8str, const XMLCh* u16str, bool escapedEnter = true) {
+    if (nullptr == u16str) {
+        return;
+    }
+    char tmpChar;
+    for (size_t i = 0; u16str[i] != u'\0';) {
+        char32_t codepoint = u16str[i++];
+
+        if (codepoint >= 0xD800 && codepoint <= 0xDBFF && u16str[i] >= 0xDC00 && u16str[i] <= 0xDFFF) {
+            // Surrogate pair handling
+            codepoint = 0x10000 + ((codepoint - 0xD800) << 10) + (u16str[i++] - 0xDC00);
+        }
+        if (codepoint <= 0x7F) {
+            tmpChar = static_cast<char>(codepoint);
+            if ((codepoint >= 0x00 && codepoint <= 0x08)
+                || (codepoint >= 0x0B && codepoint <= 0x0C)
+                || (codepoint >= 0x0E && codepoint <= 0x1F)
+                || codepoint == 0x7F) {
+                // controle charactor, ignore, \\x00-\\x08\\x0b-\\x0c\\x0e-\\x1f\\x7f
+                continue;
+            }
+            switch (tmpChar) {
+            // handle escaped chars
+            case '\"':
+                utf8str += "\\\"";
+                break;
+            case '\\':
+                utf8str += "\\\\";
+                break;
+            default:
+                if (escapedEnter) {
+                    // handle escaped chars
+                    switch (tmpChar) {
+                        case '\n': // "\n" --> "\\n"
+                            utf8str += "\\n";
+                            break;
+                        case '\r': // "\r" --> "\\r"
+                            utf8str += "\\r";
+                            break;
+                        default:
+                            utf8str += tmpChar;
+                            break;
+                    }
+                } else {
+                    utf8str += tmpChar;
+                }
+                break;
+            }
+        }
+        else if (codepoint <= 0x7FF) {
+            utf8str += static_cast<char>((codepoint >> 6) | 0xC0);
+            utf8str += static_cast<char>((codepoint & 0x3F) | 0x80);
+        }
+        else if (codepoint <= 0xFFFF) {
+            utf8str += static_cast<char>((codepoint >> 12) | 0xE0);
+            utf8str += static_cast<char>(((codepoint >> 6) & 0x3F) | 0x80);
+            utf8str += static_cast<char>((codepoint & 0x3F) | 0x80);
+        }
+        else if (codepoint <= 0x10FFFF) {
+            utf8str += static_cast<char>((codepoint >> 18) | 0xF0);
+            utf8str += static_cast<char>(((codepoint >> 12) & 0x3F) | 0x80);
+            utf8str += static_cast<char>(((codepoint >> 6) & 0x3F) | 0x80);
+            utf8str += static_cast<char>((codepoint & 0x3F) | 0x80);
+        }
+    }
+}
 
 
 // ---------------------------------------------------------------------------
@@ -47,16 +118,33 @@ MemParseHandlers::~MemParseHandlers()
 {
 }
 
-
 // ---------------------------------------------------------------------------
 //  MemParseHandlers: Implementation of the SAX DocumentHandler interface
 // ---------------------------------------------------------------------------
 void MemParseHandlers::startElement(const   XMLCh* const    /* name */
                                     ,       AttributeList&  attributes)
 {
+    // throw UserInterruption();
     fElementCount++;
     fAttrCount += attributes.getLength();
 }
+void MemParseHandlers::endElement(const XMLCh* const name)
+{
+}
+void MemParseHandlers::endElement2(const XMLCh* const name, const HandlerExtraInfo& extraInfo)
+{
+    string utf8str;
+    XMLChToUtf8(utf8str, name);
+    std::cout << "MemParseHandlers::endElement: tag: " << utf8str << std::endl;
+
+    // XMLCh fTotalSizeOfContentHasRead = name[0];
+    XMLCh fTotalSizeOfContentHasRead = extraInfo.mTotalSizeOfContentHasRead;
+    std::cout << "MemParseHandlers::endElement: " << fTotalSizeOfContentHasRead << std::endl;
+    // std::cout << "MemParseHandlers::endElement: " << gXMLInMemBuf[fTotalSizeOfContentHasRead] << std::endl;
+    printf("MemParseHandlers::endElement: %s\n", &gXMLInMemBuf[fTotalSizeOfContentHasRead]);
+
+}
+
 
 void MemParseHandlers::characters(  const   XMLCh* const    /* chars */
                                     , const XMLSize_t    length)
diff --git a/samples/src/MemParse/MemParseHandlers.hpp b/samples/src/MemParse/MemParseHandlers.hpp
index 741189a45..8f834c496 100644
--- a/samples/src/MemParse/MemParseHandlers.hpp
+++ b/samples/src/MemParse/MemParseHandlers.hpp
@@ -32,6 +32,11 @@ class AttributeList;
 XERCES_CPP_NAMESPACE_END
 
 
+class UserInterruption {
+public:
+    int mCode = 99;
+};
+
 class MemParseHandlers : public HandlerBase
 {
 public:
@@ -70,6 +75,9 @@ public:
     //  Handlers for the SAX DocumentHandler interface
     // -----------------------------------------------------------------------
     void startElement(const XMLCh* const name, AttributeList& attributes);
+    void endElement(const XMLCh* const name) override;
+    void endElement2(const XMLCh* const name, const HandlerExtraInfo&) override;
+
     void characters(const XMLCh* const chars, const XMLSize_t length);
     void ignorableWhitespace(const XMLCh* const chars, const XMLSize_t length);
     void resetDocument();
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index a168db125..382bd50af 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1276,6 +1276,12 @@ if(XERCES_USE_NETACCESSOR_CURL)
   target_include_directories(xerces-c SYSTEM PRIVATE ${CURL_INCLUDE_DIRS})
 endif()
 
+# start by bdg, static lib
+add_library(xerces-c-static STATIC
+  ${libxerces_c_SOURCES}
+  ${libxerces_c_RESOURCES})
+# end by bdg, static lib
+
 if(MSVC)
   # Add configuration-specific library name to resource file.
   target_compile_definitions(xerces-c PRIVATE "XERCES_DLL_NAME=\"$<TARGET_FILE_NAME:xerces-c>\\0\"")
diff --git a/src/xercesc/framework/XMLElementDecl.hpp b/src/xercesc/framework/XMLElementDecl.hpp
index 450136236..3fe641d4e 100644
--- a/src/xercesc/framework/XMLElementDecl.hpp
+++ b/src/xercesc/framework/XMLElementDecl.hpp
@@ -102,7 +102,9 @@ class XMLPARSER_EXPORT XMLElementDecl : public XSerializable, public XMemory
     static const unsigned int   fgPCDataElemId;
     static const XMLCh          fgPCDataElemName[];
 
-
+    // start bisheng, add by bdg
+    size_t fTotalSizeOfContentHasRead;
+    // end bisheng
 
     // -----------------------------------------------------------------------
     //  Destructor
diff --git a/src/xercesc/internal/IGXMLScanner.cpp b/src/xercesc/internal/IGXMLScanner.cpp
index 912ec0c62..0330768ee 100644
--- a/src/xercesc/internal/IGXMLScanner.cpp
+++ b/src/xercesc/internal/IGXMLScanner.cpp
@@ -44,6 +44,7 @@
 #include <xercesc/validators/schema/identity/IdentityConstraintHandler.hpp>
 #include <xercesc/validators/schema/identity/IC_Selector.hpp>
 #include <xercesc/util/OutOfMemoryException.hpp>
+#include <iostream>
 
 XERCES_CPP_NAMESPACE_BEGIN
 
@@ -926,6 +927,9 @@ bool IGXMLScanner::scanContent()
 
 void IGXMLScanner::scanEndTag(bool& gotData)
 {
+    std::cout << "IGXMLScanner::scanEndTag-1, fReaderMgr.printBuf:" << std::endl;
+    fReaderMgr.printBuf();
+
     //  Assume we will still have data until proven otherwise. It will only
     //  ever be false if this is the end of the root element.
     gotData = true;
@@ -1174,6 +1178,12 @@ void IGXMLScanner::scanEndTag(bool& gotData)
         else {
             fPrefixBuf.set(topElem->fThisElement->getElementName()->getPrefix());
         }
+        // start bisheng, add by bdg
+        // std::cout << "IGXMLScanner::scanEndTag-2, fReaderMgr.printBuf:" << std::endl;
+        fReaderMgr.printBuf();
+        size_t total = fReaderMgr.getTotalSizeOfContentHasRead();
+        const_cast<ElemStack::StackElem*>(topElem)->fThisElement->fTotalSizeOfContentHasRead = total;
+        // end bisheng
         fDocHandler->endElement
         (
             *topElem->fThisElement
@@ -2089,6 +2099,12 @@ bool IGXMLScanner::scanStartTag(bool& gotData)
     //  any prefix since its just one big name if we are not doing namespaces.
     if (fDocHandler)
     {
+        // start bisheng, add by bdg
+        // std::cout << "IGXMLScanner::scanStartTag-startElement, fReaderMgr.printBuf:" << std::endl;
+        fReaderMgr.printBuf();
+        size_t total = fReaderMgr.getTotalSizeOfContentHasRead();
+        elemDecl->fTotalSizeOfContentHasRead = total;
+        // end bisheng
         fDocHandler->startElement
         (
             *elemDecl
@@ -2631,6 +2647,12 @@ bool IGXMLScanner::scanStartTagNS(bool& gotData)
     // If we have a document handler, then tell it about this start tag
     if (fDocHandler)
     {
+        // start bisheng, add by bdg
+        // std::cout << "IGXMLScanner::scanStartTagNS-startElement, fReaderMgr.printBuf:" << std::endl;
+        fReaderMgr.printBuf();
+        size_t total = fReaderMgr.getTotalSizeOfContentHasRead();
+        elemDecl->fTotalSizeOfContentHasRead = total;
+        // end bisheng
         fDocHandler->startElement
         (
             *elemDecl
@@ -2754,6 +2776,12 @@ bool IGXMLScanner::scanStartTagNS(bool& gotData)
         // If we have a doc handler, tell it about the end tag
         if (fDocHandler)
         {
+            // start bisheng, add by bdg
+            // std::cout << "IGXMLScanner::scanStartTag-3, fReaderMgr.printBuf:" << std::endl;
+            fReaderMgr.printBuf();
+            size_t total = fReaderMgr.getTotalSizeOfContentHasRead();
+            elemDecl->fTotalSizeOfContentHasRead = total;
+            // end bisheng
             fDocHandler->endElement
             (
                 *elemDecl
diff --git a/src/xercesc/internal/ReaderMgr.cpp b/src/xercesc/internal/ReaderMgr.cpp
index d14483e3c..fa2564683 100644
--- a/src/xercesc/internal/ReaderMgr.cpp
+++ b/src/xercesc/internal/ReaderMgr.cpp
@@ -42,6 +42,7 @@
 #include <xercesc/internal/ReaderMgr.hpp>
 #include <xercesc/util/OutOfMemoryException.hpp>
 #include <xercesc/util/XMLResourceIdentifier.hpp>
+#include <iostream>
 
 XERCES_CPP_NAMESPACE_BEGIN
 
@@ -171,6 +172,25 @@ XMLCh ReaderMgr::peekNextChar()
     return chRet;
 }
 
+// start bisheng, by bdg
+void ReaderMgr::printBuf()
+{
+    if (!fCurReader) {
+        std::cerr << "Err: ReaderMgr::printBuf, fCurReader is null." << std::endl;
+        return;
+    }
+    fCurReader->printBuf();
+}
+
+size_t ReaderMgr::getTotalSizeOfContentHasRead()
+{
+    if (!fCurReader) {
+        std::cerr << "Err: ReaderMgr::getTotalSizeOfContentHasRead, fCurReader is null." << std::endl;
+        return 0;
+    }
+    return fCurReader->getTotalSizeOfContentHasRead();
+}
+// end bisheng
 
 bool ReaderMgr::skippedChar(const XMLCh toCheck)
 {
diff --git a/src/xercesc/internal/ReaderMgr.hpp b/src/xercesc/internal/ReaderMgr.hpp
index f63b2194e..d43376e08 100644
--- a/src/xercesc/internal/ReaderMgr.hpp
+++ b/src/xercesc/internal/ReaderMgr.hpp
@@ -194,6 +194,10 @@ public :
     virtual XMLFileLoc getLineNumber() const;
     virtual XMLFileLoc getColumnNumber() const;
 
+    // start bisheng, add by bdg
+    virtual void printBuf();
+    virtual size_t getTotalSizeOfContentHasRead();
+    // end bisheng
 
 private :
     // -----------------------------------------------------------------------
diff --git a/src/xercesc/internal/XMLReader.cpp b/src/xercesc/internal/XMLReader.cpp
index 405474a02..f3ada6dd3 100644
--- a/src/xercesc/internal/XMLReader.cpp
+++ b/src/xercesc/internal/XMLReader.cpp
@@ -31,6 +31,7 @@
 #include <xercesc/util/XMLEBCDICTranscoder.hpp>
 #include <xercesc/util/XMLString.hpp>
 #include <xercesc/util/Janitor.hpp>
+#include <iostream>
 
 XERCES_CPP_NAMESPACE_BEGIN
 
@@ -121,6 +122,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
     , fTranscoder(0)
     , fType(type)
     , fMemoryManager(manager)
+    // start bisheng, add by bdg
+    , fTotalReadBytes(0)
+    // end bisheng
 {
     setXMLVersion(version);
 
@@ -204,6 +208,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
     , fTranscoder(0)
     , fType(type)
     , fMemoryManager(manager)
+    // start bisheng, add by bdg
+    , fTotalReadBytes(0)
+    // end bisheng
 {
     setXMLVersion(version);
 
@@ -235,6 +242,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
                  ((fRawByteBuf[0] == 0xFF) && (fRawByteBuf[1] == 0xFE) && (fRawByteBuf[2] == 0x00) && (fRawByteBuf[3] == 0x00)))  )
             {
                 fRawBufIndex += 4;
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(4);
+                // end bisheng
             }
             break;
         }
@@ -249,6 +259,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
                                             , XMLRecognizer::fgUTF8BOMLen) == 0)
             {
                 fRawBufIndex += XMLRecognizer::fgUTF8BOMLen;
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(XMLRecognizer::fgUTF8BOMLen);
+                // end bisheng
             }
             break;
         }
@@ -262,6 +275,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
             if ((*asUTF16 == chUnicodeMarker) || (*asUTF16 == chSwappedUnicodeMarker))
             {
                 fRawBufIndex += sizeof(UTF16Ch);
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(sizeof(UTF16Ch));
+                // end bisheng
             }
             break;
         }
@@ -387,6 +403,9 @@ XMLReader::XMLReader(const  XMLCh* const          pubId
     , fTranscoder(0)
     , fType(type)
     , fMemoryManager(manager)
+    // start bisheng, add by bdg
+    , fTotalReadBytes(0)
+    // end bisheng
 {
     setXMLVersion(version);
 
@@ -1462,6 +1481,40 @@ void XMLReader::checkForSwapped()
     }
 }
 
+// start bisheng, add by bdg
+void XMLReader::printBuf()
+{
+    if (!fStream) {
+        std::cerr << " ======= Err: fStream is null." << std::endl;
+        return;
+    }
+    // XMLReader::fCharBuf[fCharIndex]  fCharsAvail?
+    // XMLReader::fCharSizeBuf
+
+    size_t numCharLeftNotRead = 0;
+    for (size_t idx = fCharIndex; idx < fCharsAvail; idx++) {
+        numCharLeftNotRead += fCharSizeBuf[idx];
+    }
+    size_t total = fTotalReadBytes - numCharLeftNotRead;
+    // std::cerr << " ======= XMLReader::printBuf: total: " << total << std::endl;
+    fStream->printBuf(total);
+}
+
+size_t XMLReader::getTotalSizeOfContentHasRead() {
+    if (!fStream) {
+        std::cerr << " ======= Err: fStream is null." << std::endl;
+        return 0;
+    }
+
+    size_t numCharLeftNotRead = 0;
+    for (size_t idx = fCharIndex; idx < fCharsAvail; idx++) {
+        numCharLeftNotRead += fCharSizeBuf[idx];
+    }
+    size_t total = fTotalReadBytes - numCharLeftNotRead;
+
+    return total;
+}
+// end bisheng
 
 //
 //  This is called from the constructor when the encoding is not forced.
@@ -1498,6 +1551,9 @@ void XMLReader::doInitDecode()
                 if (fRawBufIndex + sizeof(UCS4Ch) > fRawBytesAvail) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1515,6 +1571,9 @@ void XMLReader::doInitDecode()
                 if (fCharsAvail == kCharBufSize - 1) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1530,6 +1589,9 @@ void XMLReader::doInitDecode()
                 // Get out the current 4 byte value and inc our raw buf index
                 UCS4Ch curVal = *asUCS++;
                 fRawBufIndex += sizeof(UCS4Ch);
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(sizeof(UCS4Ch));
+                // end bisheng
 
                 // Swap if that is required for this machine
                 if (fSwapped)
@@ -1540,6 +1602,10 @@ void XMLReader::doInitDecode()
                 {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
+
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1579,6 +1645,9 @@ void XMLReader::doInitDecode()
                                             , XMLRecognizer::fgUTF8BOMLen) == 0)
             {
                 fRawBufIndex += XMLRecognizer::fgUTF8BOMLen;
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(XMLRecognizer::fgUTF8BOMLen);
+                // end bisheng
                 asChars      += XMLRecognizer::fgUTF8BOMLen;
             }
 
@@ -1602,12 +1671,18 @@ void XMLReader::doInitDecode()
             {
                 const char curCh = *asChars++;
                 fRawBufIndex++;
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(1);
+                // end bisheng
 
                 // Make sure we don't exhaust the limited prolog buffer size.
                 // Leave room for a space added at the end of this function.
                 if (fCharsAvail == kCharBufSize - 1) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1636,6 +1711,10 @@ void XMLReader::doInitDecode()
                 {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
+
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1667,6 +1746,9 @@ void XMLReader::doInitDecode()
             if ((*asUTF16 == chUnicodeMarker) || (*asUTF16 == chSwappedUnicodeMarker))
             {
                 fRawBufIndex += sizeof(UTF16Ch);
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(sizeof(UTF16Ch));
+                // end bisheng
                 asUTF16++;
                 postBOMIndex = fRawBufIndex;
             }
@@ -1677,6 +1759,9 @@ void XMLReader::doInitDecode()
             if (fRawBytesAvail - fRawBufIndex < XMLRecognizer::fgUTF16PreLen)
             {
                 fRawBufIndex = postBOMIndex;
+                // start bisheng, add by bdg, bisheng FIXME
+                // increaseTotalReadBytes(postBOMIndex);
+                // end bisheng
                 break;
             }
 
@@ -1689,6 +1774,9 @@ void XMLReader::doInitDecode()
                 if (memcmp(asUTF16, XMLRecognizer::fgUTF16BPre, XMLRecognizer::fgUTF16PreLen))
                 {
                     fRawBufIndex = postBOMIndex;
+                    // start bisheng, add by bdg, bisheng FIXME
+                    // increaseTotalReadBytes(postBOMIndex);
+                    // end bisheng
                     break;
                 }
             }
@@ -1697,6 +1785,9 @@ void XMLReader::doInitDecode()
                 if (memcmp(asUTF16, XMLRecognizer::fgUTF16LPre, XMLRecognizer::fgUTF16PreLen))
                 {
                     fRawBufIndex = postBOMIndex;
+                    // start bisheng, add by bdg, bisheng FIXME
+                    // increaseTotalReadBytes(postBOMIndex);
+                    // end bisheng
                     break;
                 }
             }
@@ -1707,6 +1798,9 @@ void XMLReader::doInitDecode()
                 if (fRawBufIndex + sizeof(UTF16Ch) > fRawBytesAvail) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1724,6 +1818,9 @@ void XMLReader::doInitDecode()
                 if (fCharsAvail == kCharBufSize - 1) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1739,6 +1836,9 @@ void XMLReader::doInitDecode()
                 // Get out the current 2 byte value
                 UTF16Ch curVal = *asUTF16++;
                 fRawBufIndex += sizeof(UTF16Ch);
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(sizeof(UTF16Ch));
+                // end bisheng
 
                 // Swap if that is required for this machine
                 if (fSwapped)
@@ -1770,12 +1870,18 @@ void XMLReader::doInitDecode()
                 // Transcode one char from the source
                 const XMLCh chCur = XMLEBCDICTranscoder::xlatThisOne(*srcPtr++);
                 fRawBufIndex++;
+                // start bisheng, add by bdg
+                increaseTotalReadBytes(1);
+                // end bisheng
 
                 // Make sure we don't exhaust the limited prolog buffer size.
                 // Leave room for a space added at the end of this function.
                 if (fCharsAvail == kCharBufSize - 1) {
                     fCharsAvail = 0;
                     fRawBufIndex = 0;
+                    // start bisheng, add by bdg
+                    resetTotalReadBytes();
+                    // end bisheng
                     fMemoryManager->deallocate(fPublicId);
                     fMemoryManager->deallocate(fEncodingStr);
                     ArrayJanitor<XMLCh> janValue(fSystemId, fMemoryManager);
@@ -1935,8 +2041,12 @@ XMLReader::xcodeMoreChars(          XMLCh* const            bufToFill
 
         if (bytesEaten == 0)
             needMode = true;
-        else
+        else {
             fRawBufIndex += bytesEaten;
+            // start bisheng, add by bdg
+            increaseTotalReadBytes(bytesEaten);
+            // end bisheng
+        }
     }
 
     return charsDone;
diff --git a/src/xercesc/internal/XMLReader.hpp b/src/xercesc/internal/XMLReader.hpp
index 966ca224d..5b1d95d38 100644
--- a/src/xercesc/internal/XMLReader.hpp
+++ b/src/xercesc/internal/XMLReader.hpp
@@ -467,6 +467,18 @@ private:
     bool                        fNEL;
     XMLVersion                  fXMLVersion;
     MemoryManager*              fMemoryManager;
+    // start bisheng, add by bdg
+    XMLSize_t                   fTotalReadBytes;
+    inline void increaseTotalReadBytes(size_t addNum) {
+        fTotalReadBytes = fTotalReadBytes + addNum;
+    }
+    inline void resetTotalReadBytes() {
+        fTotalReadBytes = 0;
+    }
+public:
+    void printBuf();
+    size_t getTotalSizeOfContentHasRead();
+    // end bisheng
 };
 
 
diff --git a/src/xercesc/parsers/SAXParser.cpp b/src/xercesc/parsers/SAXParser.cpp
index 0c80a1356..b0d0387fd 100644
--- a/src/xercesc/parsers/SAXParser.cpp
+++ b/src/xercesc/parsers/SAXParser.cpp
@@ -853,6 +853,12 @@ void SAXParser::endElement( const   XMLElementDecl& elemDecl
                             , const bool            isRoot
                             , const XMLCh* const    elemPrefix)
 {
+// start bisheng, modify by bdg
+#if 1
+    // start bisheng, add by bdg
+    HandlerExtraInfo extraInfo = {elemDecl.fTotalSizeOfContentHasRead};
+    // end bisheng
+
     // Just map to the SAX document handler
     if (fDocHandler) {
         if (fScanner->getDoNamespaces()) {
@@ -863,16 +869,52 @@ void SAXParser::endElement( const   XMLElementDecl& elemDecl
                 fElemQNameBuf.append(chColon);
                 fElemQNameBuf.append(elemDecl.getBaseName());
                 fDocHandler->endElement(fElemQNameBuf.getRawBuffer());
+                // start bisheng, add by bdg
+                fDocHandler->endElement2(fElemQNameBuf.getRawBuffer(), extraInfo);
+                // end bisheng
             }
             else {
                 fDocHandler->endElement(elemDecl.getBaseName());
+                // start bisheng, add by bdg
+                fDocHandler->endElement2(elemDecl.getBaseName(), extraInfo);
+                // end bisheng
             }
         }
-        else
+        else {
             fDocHandler->endElement(elemDecl.getFullName());
-
+            // start bisheng, add by bdg
+            fDocHandler->endElement2(elemDecl.getFullName(), extraInfo);
+            // end bisheng
+        }
     }
+#else
+    // Just map to the SAX document handler
+    if (fDocHandler) {
+        const XMLCh* ptrXmlCh = 0;
+        if (fScanner->getDoNamespaces()) {
 
+            if (elemPrefix && *elemPrefix) {
+
+                fElemQNameBuf.set(elemPrefix);
+                fElemQNameBuf.append(chColon);
+                // fDocHandler->endElement(fElemQNameBuf.getRawBuffer());
+                ptrXmlCh = elemDecl.getBaseName();
+            }
+            else {
+                // fDocHandler->endElement(elemDecl.getBaseName());
+                ptrXmlCh = elemDecl.getBaseName();
+            }
+        }
+        else {
+            // fDocHandler->endElement(elemDecl.getFullName());
+            ptrXmlCh = elemDecl.getFullName();
+        }
+
+        fDocHandler->endElement(ptrXmlCh);
+        fDocHandler->endElement2(ptrXmlCh, extraInfo);
+    }
+#endif
+// end bisheng
     //
     //  If there are any installed advanced handlers, then lets call them
     //  with this info.
@@ -968,6 +1010,9 @@ startElement(   const   XMLElementDecl&         elemDecl
     // Bump the element depth counter if not empty
     if (!isEmpty)
         fElemDepth++;
+    // start bisheng, add by bdg
+    HandlerExtraInfo extraInfo = {elemDecl.fTotalSizeOfContentHasRead};
+    // end bisheng
 
     if (fDocHandler)
     {
@@ -982,24 +1027,36 @@ startElement(   const   XMLElementDecl&         elemDecl
                 fDocHandler->startElement(fElemQNameBuf.getRawBuffer(), fAttrList);
 
                 // If its empty, send the end tag event now
-                if (isEmpty && fDocHandler)
+                if (isEmpty && fDocHandler) {
                     fDocHandler->endElement(fElemQNameBuf.getRawBuffer());
+                    // start bisheng, add by bdg
+                    fDocHandler->endElement2(fElemQNameBuf.getRawBuffer(), extraInfo);
+                    // end bisheng
+                }
             }
             else {
 
                 fDocHandler->startElement(elemDecl.getBaseName(), fAttrList);
 
                 // If its empty, send the end tag event now
-                if (isEmpty && fDocHandler)
+                if (isEmpty && fDocHandler) {
                     fDocHandler->endElement(elemDecl.getBaseName());
+                    // start bisheng, add by bdg
+                    fDocHandler->endElement2(elemDecl.getBaseName(), extraInfo);
+                    // end bisheng
+                }
             }
         }
         else {
             fDocHandler->startElement(elemDecl.getFullName(), fAttrList);
 
             // If its empty, send the end tag event now
-            if (isEmpty && fDocHandler)
+            if (isEmpty && fDocHandler) {
                 fDocHandler->endElement(elemDecl.getFullName());
+                // start bisheng, add by bdg
+                fDocHandler->endElement2(elemDecl.getFullName(), extraInfo);
+                // end bisheng
+            }
         }
     }
 
diff --git a/src/xercesc/sax/DocumentHandler.hpp b/src/xercesc/sax/DocumentHandler.hpp
index 6aecddb64..aa931d44c 100644
--- a/src/xercesc/sax/DocumentHandler.hpp
+++ b/src/xercesc/sax/DocumentHandler.hpp
@@ -56,6 +56,11 @@ class Locator;
   * @see Locator#Locator
   * @see HandlerBase#HandlerBase
   */
+// start bisheng, add by bdg
+struct SAX_EXPORT HandlerExtraInfo {
+  size_t mTotalSizeOfContentHasRead = 0;
+};
+// end bisheng
 
 class SAX_EXPORT DocumentHandler
 {
@@ -136,6 +141,10 @@ public:
     *            wrapping another exception.
     */
     virtual void endElement(const XMLCh* const name) = 0;
+    // add start bisheng, modify by bdg
+    virtual void endElement2(const XMLCh* const name,
+        const HandlerExtraInfo& extraInfo) {}
+    // end bisheng
 
   /**
     * Receive notification of ignorable whitespace in element content.
diff --git a/src/xercesc/sax/HandlerBase.hpp b/src/xercesc/sax/HandlerBase.hpp
index 12a5cbe2a..d6972c46b 100644
--- a/src/xercesc/sax/HandlerBase.hpp
+++ b/src/xercesc/sax/HandlerBase.hpp
@@ -110,6 +110,10 @@ public:
     * @see DocumentHandler#endElement
     */
     virtual void endElement(const XMLCh* const name);
+    // add start bisheng, add by bdg
+    virtual void endElement2(const XMLCh* const name,
+        const HandlerExtraInfo& extraInfo);
+    // end bisheng
 
   /**
     * Receive notification of ignorable whitespace in element content.
@@ -389,6 +393,13 @@ inline void HandlerBase::endElement(const XMLCh* const)
 {
 }
 
+// add start bisheng, add by bdg
+inline void HandlerBase::endElement2(const XMLCh* const,
+        const HandlerExtraInfo& extraInfo)
+{
+}
+// end bisheng
+
 inline void HandlerBase::error(const SAXParseException&)
 {
 }
diff --git a/src/xercesc/util/BinInputStream.hpp b/src/xercesc/util/BinInputStream.hpp
index 6023f4897..a9295b5c0 100644
--- a/src/xercesc/util/BinInputStream.hpp
+++ b/src/xercesc/util/BinInputStream.hpp
@@ -84,6 +84,9 @@ public :
      */
     virtual const XMLCh *getEncoding() const;
 
+    // start bisheng, add by bdg
+    virtual void printBuf(size_t readIndex) {}
+    // end bisheng
 protected :
     // -----------------------------------------------------------------------
     //  Hidden Constructors
diff --git a/src/xercesc/util/BinMemInputStream.cpp b/src/xercesc/util/BinMemInputStream.cpp
index 5ebdf9936..defa68c06 100644
--- a/src/xercesc/util/BinMemInputStream.cpp
+++ b/src/xercesc/util/BinMemInputStream.cpp
@@ -26,6 +26,8 @@
 #include <xercesc/util/BinMemInputStream.hpp>
 #include <xercesc/framework/MemoryManager.hpp>
 #include <string.h>
+#include <iostream>
+#include <stdio.h>
 
 XERCES_CPP_NAMESPACE_BEGIN
 
@@ -90,6 +92,15 @@ XMLSize_t BinMemInputStream::readBytes(       XMLByte* const  toFill
     return actualToRead;
 }
 
+// start bisheng, by bdg
+void BinMemInputStream::printBuf(size_t readIndex) {
+    // std::cout << "" fBuffer[readIndex]
+    // std::cout << "BinMemInputStream::printBuf-1: readIndex: " << readIndex << std::endl;
+    // std::cout << "BinMemInputStream::printBuf-2: fBuffer: " << fBuffer[readIndex] << std::endl;
+    // printf("BinMemInputStream::printBuf-3: %s\n", &fBuffer[readIndex]);
+}
+// end bisheng
+
 const XMLCh* BinMemInputStream::getContentType() const
 {
     return 0;
diff --git a/src/xercesc/util/BinMemInputStream.hpp b/src/xercesc/util/BinMemInputStream.hpp
index 1617b65d0..dbccf8f5d 100644
--- a/src/xercesc/util/BinMemInputStream.hpp
+++ b/src/xercesc/util/BinMemInputStream.hpp
@@ -75,6 +75,9 @@ public :
 
     inline XMLSize_t getSize() const;
 
+    // start bisheng, add by bdg
+    virtual void printBuf(size_t readIndex);
+    // end bisheng
 private :
     // -----------------------------------------------------------------------
     //  Unimplemented constructors and operators
